### DISCUSSION TITLE
Meaning of the offset

### DISCUSSION THREAD

**Discussion by AlbertoPresta** [InterDigitalInc/CompressAI#168] [Q&A] [answered]
Hi, 

Maybe is a stupid question, but what is the physical meaning of the value called self._offset? 
You calculated in the following way: 

self._offset = -minima

**Comment by YodaEmbedding** [comment, marked as answer]
`_offset` is set in `update()`:
https://github.com/InterDigitalInc/CompressAI/blob/216af74fa09a141eaab68dd0ff2edd89629f62dc/compressai/entropy_models/entropy_models.py#L395-L405

And is later accessed during `compress()`:
https://github.com/InterDigitalInc/CompressAI/blob/216af74fa09a141eaab68dd0ff2edd89629f62dc/compressai/entropy_models/entropy_models.py#L260-L266

...and is immediately passed into the entropy coder, which subtracts the offset from the symbol it is trying to encode:
https://github.com/InterDigitalInc/CompressAI/blob/216af74fa09a141eaab68dd0ff2edd89629f62dc/compressai/cpp_exts/rans/rans_interface.cpp#L128

---

Each CDF has associated with it a single `offset` scalar value. This offset represents the smallest symbol that is encodable by our zero-mean distribution. (For even smaller symbols outside the distribution, we use a bypass coding engine.) Thus, `offset ≤ 0`.

```
ORIGINAL VALUES (FOR ZERO-MEAN SYMBOLS):

  ▲
  │            xxx
  │          xx │ xx
  │         x   │  xx
  │        x    │    x
  │       x     │    xx
  │      x      │      xx
  │     xx      │       xx
  │    xx       │       │xx
  │xxxx         │       │ xxxx
  ├─────────────┼───────┼───────►
  │             │       │

offset          0     symbol
```

```
MEMORY:

  ▲
  │            xxx
  │          xx │ xx
  │         x   │  xx
  │        x    │    x
  │       x     │    xx
  │      x      │      xx
  │     xx      │       xx
  │    xx       │       │xx
  │xxxx         │       │ xxxx
  ├─────────────┼───────┼───────►
  │             │       │

  0        -offset    (symbol - offset)
```

Notice how the "zero mean" `symbol` might be any number -- even negative! But we want to do an indexing operation on the CDF in memory, which only accepts positive numbers. That's what `(symbol - offset)` ensures.

**Comment by AlbertoPresta** [comment]
ok thanks got it! and just to be sure at 100%, the cdf_length simply represents the length of the cdf right? i know it can be seem a stupid question, but just to understan fully the code

**Reply by YodaEmbedding** [reply]
`quantized_cdf` is a 2D array. However, since CDFs are of different sizes, we use `cdf_length[i]` to represent the number of valid elements within the 1D array `quantized_cdf[i]`. Any element `quantized_cdf[i][j]` such that `j ≥ cdf_length[i]` is considered invalid, and thus will never be read/accessed.

https://github.com/InterDigitalInc/CompressAI/blob/216af74fa09a141eaab68dd0ff2edd89629f62dc/compressai/entropy_models/entropy_models.py#L428

Related: https://github.com/InterDigitalInc/CompressAI/discussions/146#discussioncomment-2949648
